packages/txservice/src/provider.ts

```diff
@@ -461,128 +492,197 @@ export class ChainRpcProvider {
+    for (const provider of this.providers) {
+      const providerBlockNumber = provider.syncedBlockNumber;
+      provider.lag = highestBlockNumber - providerBlockNumber;
+
+      // Set synced property, log if the provider went out of sync.
+      const synced = provider.lag < PROVIDER_MAX_LAG;
+      if (!synced && provider.synced) {
+        // If the provider was previously synced but fell out of sync, debug log to notify.
+        this.logger.debug("Provider fell out of sync.", undefined, undefined, {
+          providerBlockNumber,
+          provider: provider.url,
+          lag: provider.lag,
+        });
+      }
+      provider.synced = synced;
     }
-    return true;
+
+    // We want to pick the lead provider here at random from the list of 0-lag providers to ensure that we distribute
+    // our block listener RPC calls as evenly as possible across all providers.
+    const leadProviders = this.shuffleSyncedProviders(true);
+    this.leadProvider = leadProviders[(Math.random() * leadProviders.length) | 0];
+
+    this.logger.debug("Synced provider(s).", requestContext, methodContext, {
+      highestBlockNumber,
+      leadProvider: this.leadProvider.url,
+      providers: this.providers
+        .filter((p) => p.synced)
+        .map((p) => ({
+          url: p.url,
+          blockNumber: p.syncedBlockNumber,
+          lag: p.lag,
+        })),
+    });
+  }
+
+  /**
+   * Helper method to stall, possibly until we've surpassed a specified number of blocks. Only works
+   * with block number if we're running in synchronized mode.
+   *
+   * @param numBlocks (default: 1) - the number of blocks to wait.
+   */
+  private async wait(numBlocks = 1): Promise<void> {
+    const pollPeriod = numBlocks * (this.blockPeriod ?? 2_000);
+    await delay(pollPeriod);
+  }
+
+  /**
+   * Helper method for getting tier-shuffled synced providers.
+   *
+   * @returns all in-sync providers in order of synchronicity with chain, with the lead provider
+   * in the first position and the rest shuffled by tier (lag).
+   */
+  private shuffleSyncedProviders(zeroLagOnly = false): SyncProvider[] {
+    const syncedProviders = this.providers.filter((p) => (zeroLagOnly ? p.lag === 0 : p.synced));
+    // Tiered shuffling: providers that have the same lag value (e.g. 0) will be shuffled so as to distribute RPC calls
+    // as evenly as possible across all providers; at high load, this can translate to higher efficiency (each time we
+    // execute an RPC call, we'll be hitting different providers).
+    // Shuffle isn't applied to lead provider - instead, we just guarantee that it's in the first position.
+    syncedProviders.forEach((p) => {
+      p.priority =
+        p.lag -
+        (this.leadProvider && p.url === this.leadProvider.url ? 1 : Math.random()) -
+        p.cps / this.config.maxProviderCPS -
+        p.reliability * 2 +
+        p.avgExecTime;
+    });
+    return syncedProviders.sort((a, b) => a.priority - b.priority);
```

Comment:
do we want to do something similar with the unsynced providers being used at the back of the list, or simply *only* include the synced providers?

Reply:
sync is all relative here. technically speaking I could be running with a single provider that's 300 blocks behind but i wouldn't know it, bc I have no other provider to compare it to.

as it is, lag is the most influential factor here by far. being 10 blocks behind another provider basically guarantees that provider will be chosen after the one that's ahead.

Buuut if you'd like, I could choose to do what I did with fallback subg. Make the ..MAX_LAG value effectively a line in the sand, and have all providers that are lagging < than that in front and concatenate all laggy providers.

So we guarantee, regardless of any metrics, that we always try those "in-sync" providers first